pkg/kubelet/dockertools/images.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockertools
+
+import (
+	"fmt"
+
+	"github.com/fsouza/go-dockerclient"
+	"github.com/golang/glog"
+
+	runtime "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// imageStatsProvider exposes stats about all images currently available.
+type imageStatsProvider struct {
+	// Docker remote API client
+	c DockerInterface
+}
+
+func (isp *imageStatsProvider) ImageStats() (*runtime.ImageStats, error) {
+	images, err := isp.c.ListImages(docker.ListImagesOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list docker images - %v", err)
+	}
+	// A map of all the image layers to its corresponding size.
+	imageMap := sets.NewString()
+	ret := &runtime.ImageStats{}
+	for _, image := range images {
+		// Get information about the various layers of each docker image.
+		history, err := isp.c.ImageHistory(image.ID)
+		if err != nil {
+			glog.V(2).Infof("failed to get history of docker image %v - %v", image, err)
+			continue
+		}
+		// Store size information of each layer.
+		for _, layer := range history {
+			// Skip empty layers.
+			if layer.Size == 0 {
+				glog.V(10).Infof("skipping image layer %v with size 0", layer)
+				continue
+			}
+			key := layer.ID
+			// Some of the layers are empty.
+			// We are hoping that these layers are unique to each image.
+			// Still keying with the CreatedBy field to be safe.
+			if key == "" || key == "<missing>" {
+				key = key + layer.CreatedBy
+			}
+			if !imageMap.Has(key) {
+				ret.TotalStorageBytes += uint64(layer.Size)
+			}
+			imageMap.Insert(key)
```

Comment:
Hmm. `history` contains all the layers for an image. Let's look at an example:

```shell
$ docker history bef5c7a893fd
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
bef5c7a893fd        4 days ago          /bin/sh -c #(nop) ENTRYPOINT ["/usr/bin/cadvi   0 B                 
5714946567e8        4 days ago          /bin/sh -c #(nop) EXPOSE 8080/tcp               0 B                 
d5092c6dfc49        4 days ago          /bin/sh -c #(nop) ADD file:5047bc1f57482bfdc9   23.75 MB            
fb1790e81294        4 days ago          /bin/sh -c apk add --update ca-certificates d   19.99 MB            
3d239f12c85f        4 weeks ago         /bin/sh -c #(nop) ENV GLIBC_VERSION=2.23-r1     0 B                 
86fbae24586c        4 weeks ago         /bin/sh -c #(nop) MAINTAINER dengnan@google.c   0 B                 
342c0650e86b        7 weeks ago         /bin/sh -c #(nop) ADD file:cda4b589f22e7984e3   5.258 MB 
```

For this example, history would be a list of layer elements where each line in the output above matches a layer.
The key for each layer will be mostly unique, expect for the `<missing>` case, which is handled by using the `Created By` field. That field is expected to contain the command that was run for that layer and hence it must be unique.

Am I misreading my code?